pkg/dag/packages.go

```diff
@@ -0,0 +1,222 @@
+				}
+			}
+
+			for i := range c.Subpackages {
+				subpkg := c.Subpackages[i]
+				if subpkg.Name == "" {
+					log.Fatalf("empty subpackage name at index %d for package %q", i, c.Package.Name)
+				}
+				if err := pkgs.addConfiguration(subpkg.Name, c); err != nil {
+					return err
+				}
+
+				for _, prov := range subpkg.Dependencies.Provides {
+					name, version := packageNameFromProvides(prov)
+					if err := pkgs.addNameVersionDescription(name, version, fmt.Sprintf("PROVIDED BY %s", c.Package.Name), c.Path); err != nil {
+						return err
+					}
+				}
+
+				// TODO: resolve deps via `uses` for subpackage pipelines.
+			}
+			// Resolve all `uses` used by the pipeline. This updates the set of
+			// .environment.contents.packages so the next block can include those as build deps.
+			pctx := &build.PipelineContext{
+				Context: &build.Context{
+					Configuration: *c.Configuration,
+				},
+				Package: &c.Package,
+			}
+			for i := range c.Pipeline {
+				s := c.Pipeline[i]
+				if err := s.ApplyNeeds(pctx); err != nil {
+					return fmt.Errorf("unable to resolve needs for package %s: %w", name, err)
+				}
+				c.Environment.Contents.Packages = pctx.Context.Configuration.Environment.Contents.Packages
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return pkgs, nil
+}
+
+// Config returns the Melange configuration for the package, provides or subpackage with the given name,
+// if the package is present in the Graph. If it's not present, Config returns
+// nil. Providing the name of a subpackage will return the configuration of the
+// subpackage's origin package.
+func (p Packages) Config(name string) []*Configuration {
+	if p.configs == nil {
+		// this would be unexpected
+		return nil
+	}
+
+	if c, ok := p.configs[name]; ok {
+		list := make([]*Configuration, 0, len(c))
+		list = append(list, c...)
+		return list
+	}
+
+	return nil
+}
+
+// PkgInfo returns the buildp.Package struct
+func (p Packages) PkgInfo(pkgName string) (*build.Package, error) {
+	config := p.Config(pkgName)
+	if len(config) == 0 {
+		log.Println("no config for package:", pkgName)
+		return nil, nil
+	}
+	c := config[0]
+	if pkgName != c.Package.Name {
+		return nil, nil
+	}
+	return &c.Package, nil
+}
+
+// Packages returns a slice of the names of all packages, sorted alphabetically.
+func (p Packages) Packages() []string {
+	allPackages := make([]string, 0)
+	for name := range p.packages {
+		allPackages = append(allPackages, name)
+	}
+
+	// sort for deterministic output
+	sort.Strings(allPackages)
+	return allPackages
+}
+
+// Sub returns a new Graph whose members are the named packages or provides that are listed.
+// If a listed element is a provides, automatically includes the package that provides it.
+// If a listed element is a subpackage, automatically includes the package that contains it.
+// If a listed element does not exist, returns an error.
+func (p Packages) Sub(names ...string) (*Packages, error) {
```

Comment:
I like this function. I stumbled over the name for a second, debating if this referred to subpackages or subgraphs. (My guess is the latter?) Could we update the name for clarity?

Reply:
Okay got it. The docs say `Sub` returns a new `Graph` — but it looks like it's returning a new `*Packages`. And it looks like this populates the `Packages` object with more than just "subpackages". Maybe `SubSelection` or something would be better here?

Reply:
Updated. I really missed that one. Thanks for that @luhring